chia/data_layer/s3_plugin_service.py

```diff
@@ -160,20 +162,48 @@ async def download(self, request: web.Request) -> web.Response:
             if not should_download:
                 return web.json_response({"downloaded": False})
 
-            bucket = parse_result.netloc
+            bucket_str = parse_result.netloc
+            my_bucket = self.boto_resource.Bucket(bucket_str)
             target_filename = self.server_files_path.joinpath(filename)
             # Create folder for parent directory
             target_filename.parent.mkdir(parents=True, exist_ok=True)
             log.info(f"downloading {url} to {target_filename}...")
             with concurrent.futures.ThreadPoolExecutor() as pool:
                 await asyncio.get_running_loop().run_in_executor(
-                    pool, functools.partial(self.boto_client.download_file, bucket, filename, str(target_filename))
+                    pool, functools.partial(my_bucket.download_file, filename, str(target_filename))
                 )
         except Exception as e:
             log.error(f"failed parsing request {request} {type(e).__name__} {e}")
             return web.json_response({"downloaded": False})
         return web.json_response({"downloaded": True})
 
+    async def add_missing_files(self, request: web.Request) -> web.Response:
+        try:
+            data = await request.json()
+            store_id = bytes32.from_hexstr(data["store_id"])
+            bucket_str = self.get_bucket(store_id)
+            files = json.loads(data["files"])
+            my_bucket = self.boto_resource.Bucket(bucket_str)
+            existing_file_list = []
+            for my_bucket_object in my_bucket.objects.all():
+                existing_file_list.append(my_bucket_object.key)
+            try:
+                for file in files:
+                    file_path = Path(file)
+                    if file_path.name not in existing_file_list:
+                        with concurrent.futures.ThreadPoolExecutor() as pool:
+                            await asyncio.get_running_loop().run_in_executor(
+                                pool,
+                                functools.partial(my_bucket.upload_file, file_path, file_path.name),
+                            )
+            except ClientError as e:
+                log.error(f"failed uploading file to aws {e}")
+                return web.json_response({"uploaded": False})
+        except Exception as e:
+            log.error(f"failed handling request {request} {e}")
```

Comment:
`{type(e).__name__} {e}` is a better pattern since it prints out the name of the Exception (like KeyError, ValueError) etc. TILFK (Things I Learned From Kyle)

Reply:
isn't the name of the exception printed anyway when you dump e ?

i dont mind changing if you insist but theres specific log lines for each exception case i dont think this information is interesting to add

Reply:
The string conversion of the exception is just the message, not the type.  In many cases this is sufficient but in other cases you get silly stuff like `3`, such as for a `KeyError`.

This case is specifically `except Exception as e:` so I'm not sure we can say there is an exception for each case.

As a detail, I suggested retaining the `:` but Earle overlooked that.  `:]`  `f"our whatever message: {type(e).__name__}: {e}`.  But also, this topic needs general discussion resulting in a generally encouraged form and if we don't just use `logging.exception()` then we probably ought to make our own helper to aid in clarity, consistency, and any future changes.  Certainly, doing this is a future task, not something for this PR to wait for.